add client method for /info endpoint

seebs opened the pull request:
This unpacks additional server data which is added to the
server by a corresponding PR, and adds a direct call for
the /info endpoint as a convenience feature. This is intended
to be helpful for benchmarking-type functionality.

seebs commented:
that seems like a really good idea.

yuce commented on client_it_test.go:
```diff
@@ -939,6 +939,17 @@ func TestFetchStatus(t *testing.T) {
+	if err != nil {
+		t.Fatal(err)
+	}
+	if info.ShardWidth == 0 {
```
I think adding checks to other fields too would be trivial and useful.

seebs replied:
It might, but that would have to wait on the server being updated. My thought was to wait a release cycle or so after the server updates; I don't want the client to require a newer version of the server if it doesn't have to.

yuce replied:
The tests run against pilosa/master on CI, so I think that won't be a problem.

seebs replied:
It would right now, because the corresponding change isn't in yet. I'm also concerned a bit about users who might want to update the client without having updated the server yet; I don't want to break the tests in the absence of an underlying compatibility problem.

yuce replied:
The way I deal with changes on the server which are not yet merged to master is mentioning the required server-side PR in the ticket (description or a comment) and maybe mark the issue as `[DON'T MERGE]`. We can merge the PR after the corresponding server-side PR is merged. Would that be OK?

seebs replied:
I think so? Mostly my vague concern is that I'd rather not introduce a test failure for people using slightly-older Pilosa. If someone's not using this feature, and doesn't care about it, it introduces a possible failure even though go-pilosa is otherwise compatible with the older server, so far as I know?

yuce replied:
I see your point, but making the tests pass is trivial in that case, upgrading the server is adequate.  We release a new version of Pilosa roughly each month, so if we supported 2 older Pilosa releases for **client-side tests** we would have to postpone writing tests for 2 months. That means, for two months we wouldn't know whether those new features worked or not on the client-side. As a bonus, it would be much harder to write tests for all of those features at once instead of one by one (if we ever remember writing the tests after 2 months the new features are introduced).

I think the benefits of having timely client-side tests far outweighs the benefits of  **client-side tests** being incompatible with older Pilosa servers.

seebs replied:
... That's a good point. I'll add some tests for the other fields. (Not for Hz, because I don't actually have confidence that it'll reliably have data, yet; it's guesswork.)

seebs commented:
Updated this to match changes in the server-side implementation.

yuce commented:
Great! The PR works with the corresponding Pilosa PR: https://github.com/pilosa/pilosa/pull/1859

yuce commented:
@seebs Since https://github.com/pilosa/pilosa/pull/1859 was merged, is it OK to merge this?

seebs commented:
I'm not sure how we manage "releases"; I'd say either now, or after the next release of the server.

jaffee commented:
What we've typically said is the latest release of a client (or the pdk or
whatever) should work with the latest release of Pilosa. Master should
generally be compatible with master, though they may be incompatible for
short periods.

On Wed, Mar 27, 2019 at 11:53 AM seebs <notifications@github.com> wrote:

> I'm not sure how we manage "releases"; I'd say either now, or after the
> next release of the server.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/pilosa/go-pilosa/pull/216#issuecomment-477251209>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AA4hcyqgaE2ytoHh6BD9GcF10RK8URTNks5va6H3gaJpZM4a32Jw>
> .
>


-- 

<http://www.pilosa.com/>

Matt Jaffee

Slothware Engineer

(302) 893-0193 <(302)%893-0193> | jaffee@pilosa.com

pilosa.com <https://www.pilosa.com/> | @mattjaffee
<https://twitter.com/mattjaffee> | LinkedIn
<https://www.linkedin.com/company/pilosa> | Facebook
<https://business.facebook.com/pilosaco> | GitHub
<https://github.com/pilosa/pilosa>

yuce commented:
The changes in this PR don't break any compatibility with current or past versions and the master. Let's merge it unless there is a compelling reason not to.

seebs commented:
I should probably fix the test failures, at least. And the linter failures. I've asked CircleCI to rerun in case the updates to pilosa-master  fix any of it, but I think the lint is probably real.

seebs commented:
Okay, tests seem to be passing, and I fixed the output from a test, dropped a bogus comment, and ran gofmt. The old linter claim about goimports needing to be run seems to be wrong now, or possibly goimports implied gofmt also. Rebased and repushed, if tests work I'll merge.
